iml-manager-cli/src/server.rs

```diff
@@ -3,29 +3,155 @@
+    Add(AddHosts),
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+struct TestHostConfig<'a> {
+    address: &'a str,
+    auth_type: String,
+}
+
+#[derive(serde::Serialize, serde::Deserialize)]
+struct AgentConfig<'a> {
+    address: &'a str,
+    auth_type: String,
+    server_profile: String,
+}
+
+#[derive(serde::Serialize)]
+struct HostProfileConfig<'a> {
+    host: u32,
+    profile: &'a str,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+struct CommandAndHost {
+    command: Command,
+    host: Host,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+struct CommandAndHostWrapper {
+    command_and_host: Option<CommandAndHost>,
+    error: Option<String>,
+    traceback: Option<String>,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+struct HostProfileCmdWrapper {
+    commands: Vec<Command>,
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct Objects<T> {
+    objects: Vec<T>,
+}
+
+fn filter_known_hosts<'a, 'b>(
+    hostlist: &'b BTreeSet<String>,
+    api_hosts: &'a Vec<Host>,
+) -> Vec<&'a Host> {
+    api_hosts
+        .iter()
+        .filter(move |x| hostlist.contains(&x.fqdn) || hostlist.contains(&x.nodename))
+        .collect()
+}
+
+fn is_profile_valid(x: &HostProfile, profile_name: &str) -> bool {
+    let checks = x.profiles.get(profile_name);
+
+    match checks {
+        Some(xs) => xs.iter().all(|y| y.pass),
+        None => false,
+    }
+}
+
+async fn wait_till_agent_starts(
+    hosts: &Vec<Host>,
+    profile_name: &str,
+) -> Result<(), ImlManagerCliError> {
+    loop {
+        let host_ids: Vec<_> = hosts
+            .iter()
+            .map(|x| ["id__in".into(), x.id.to_string()])
+            .chain(iter::once(["limit".into(), "0".into()]))
+            .collect();
+
+        let ApiList { mut objects, .. }: ApiList<HostProfileWrapper> =
+            get(HostProfile::endpoint_name(), host_ids).await?;
+
+        tracing::debug!("Host Profiles {:?}", objects);
+
+        if let Some(x) = objects.iter_mut().find(|x| x.error.is_some()) {
+            return Err(ImlManagerCliError::ApiError(
+                x.error.take().unwrap().to_string(),
+            ));
+        };
+
+        let profile_checks: Vec<bool> = objects
+            .iter_mut()
+            .filter_map(|x| x.host_profiles.take())
+            .map(|x| {
+                x.profiles
+                    .get(profile_name)
+                    .map(|checks| checks.iter().all(|y| y.pass))
+                    .ok_or(ImlManagerCliError::ApiError(format!(
+                        "Profile {} not found for host {} while booting",
+                        profile_name, x.host
+                    )))
+            })
+            .collect::<Result<Vec<bool>, ImlManagerCliError>>()?;
+
+        if profile_checks.into_iter().all(std::convert::identity) {
+            return Ok(());
+        }
+
+        delay(Instant::now() + Duration::from_millis(500)).await;
+    }
```

Comment:
This fn will poll continually if the profile is truly invalid. 

The easiest thing to do at the moment would be to set a timeout on agent starting, which will exit if there are no updates after n ticks.